cmd/vm/switch_linux.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright © 2023 SUSE LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/sirupsen/logrus"
+	"github.com/songgao/packets/ethernet"
+	"github.com/songgao/water"
+	"github.com/vishvananda/netlink"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+)
+
+var (
+	debug    bool
+	tapIface string
+)
+
+const (
+	defaultTapDevice = "eth0"
+	defaultMacAddr   = "5a:94:ef:e4:0c:ee"
+	maxMTU           = 4000
+)
+
+func main() {
+	flag.BoolVar(&debug, "debug", true, "enable debug flag")
+	flag.StringVar(&tapIface, "tap-interface", defaultTapDevice, "tap interface name, eg. eth0, eth1")
+	flag.Parse()
+
+	if debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	// the FD is passed-in as an extra arg from exec.Command
+	// of the parent process. This is for the AF_VSOCK connection that
+	// is handed over from the default namespace to Rancher Desktop's
+	// network namespace, the logic behind this approach is because
+	// AF_VSOCK is affected by network namespaces, therefore we need
+	// to open it before entering a new namespace (via unshare/nsenter)
+	connFile := os.NewFile(uintptr(3), "vsock connection")
+	connFile.Close()
+
+	logrus.Debugf("using a AF_VSOCK connection file from default namespace: %v", connFile)
+
+	// this should never happen
+	if err := checkForExistingIface(tapIface); err != nil {
+		logrus.Fatal(err)
+	}
+
+	// catch user issued signals
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+
+	// try every second until we get DHCP
+	retryTicker := time.NewTicker(time.Second)
+	for {
+		ctx, cancel := context.WithCancel(context.Background())
+		select {
+		case s := <-sigChan:
+			logrus.Errorf("signal caught: %v", s)
+			cancel()
+			os.Exit(1)
```

Comment:
FWIW, if you want to let deferred things run (given that this is `main()`), one way would be:
```golang
func main() {
  exitCode := 0
  defer os.Exit(exitCode)
…
  // in this code here
  exitCode = 1
  return
```